src/openrct2/paint/tile_element/Surface.cpp

```diff
@@ -0,0 +1,1922 @@
+        {
+            boundBoxOffsetZ += 16;
+            boundBoxLength -= 16;
+        }
+
+        uint32 image_id = _terrainEdgeTunnelSpriteIds[edgeStyle][tunnelType] + (edge == EDGE_BOTTOMRIGHT ? 2 : 0);
+        sub_98197C(session, image_id, offset.x, offset.y, tunnelBounds.x, tunnelBounds.y, boundBoxLength - 1, zOffset, 0, 0,
+                   boundBoxOffsetZ, rotation);
+
+        boundBoxOffsetZ = curHeight * 16;
+        boundBoxLength  = stru_97B570[tunnelType][1] * 16;
+        boundBoxOffsetZ += word_97B590[tunnelType];
+        if (boundBoxOffsetZ == 0)
+        {
+            boundBoxOffsetZ += 16;
+            boundBoxLength -= 16;
+        }
+
+        image_id = _terrainEdgeTunnelSpriteIds[edgeStyle][tunnelType] + (edge == EDGE_BOTTOMRIGHT ? 2 : 0) + 1;
+        sub_98197C(session, image_id, offset.x, offset.y, tunnelBounds.x, tunnelBounds.y, boundBoxLength - 1, curHeight * 16,
+                   tunnelTopBoundBoxOffset.x, tunnelTopBoundBoxOffset.y, boundBoxOffsetZ, rotation);
+
+        curHeight += stru_97B570[tunnelType][0];
+
+        // TODO: Should probably be done by just keeping track of the current index
+        memmove(&tunnelArray[0], &tunnelArray[1], sizeof(tunnel_entry) * (TUNNEL_MAX_COUNT - 1));
+    }
+}
+
+/**
+ * rct2: 0x0066062C
+ *
+ * @param direction (cl)
+ * @param height (dx)
+ * @param tile_element (esi)
+ */
+void surface_paint(paint_session * session, uint8 direction, uint16 height, rct_tile_element * tileElement)
+{
+    rct_drawpixelinfo * dpi = session->Unk140E9A8;
+    session->InteractionType = VIEWPORT_INTERACTION_ITEM_TERRAIN;
+    session->DidPassSurface = true;
+    session->SurfaceElement = tileElement;
+
+    const uint16 zoomLevel = dpi->zoom_level;
+    const uint8 rotation = get_current_rotation();
+    const sint32 terrain_type = tile_element_get_terrain(tileElement);
+    const uint8 surfaceShape = viewport_surface_paint_setup_get_relative_slope(tileElement, rotation);
+
+    const LocationXY16 base = 
+    {
+        session->SpritePosition.x,
+        session->SpritePosition.y
+    };
+
+    corner_height ch = corner_heights[surfaceShape];
+
+    const uint8 height_16 = height / 16;
+
+    tile_descriptor tileDescriptors[5];
+
+    tileDescriptors[0] = 
+    {
+        tileElement,
+        terrain_type,
+        surfaceShape,
+        {
+            (uint8)(height_16 + ch.top),
+            (uint8)(height_16 + ch.right),
+            (uint8)(height_16 + ch.bottom),
+            (uint8)(height_16 + ch.left),
+        }
+    };
+
+    for (sint32 i = 0; i < 4; i++) 
+    {
+        const LocationXY16 offset = viewport_surface_paint_data[i][rotation];
+        const LocationXY16 position = 
+        { 
+            (sint16)(base.x + offset.x), 
+            (sint16)(base.y + offset.y),
+        };
+
+        tile_descriptor& descriptor = tileDescriptors[i + 1];
+        descriptor.tile_element = nullptr;
+        if (position.x > 0x2000 || position.y > 0x2000)
```

Comment:
@Gymnasiast Does this have something to do with map size limits?

Reply:
Also this case is already guarded by map_get_surface_element_at -> map_get_first_element_at which would return NULL. The loop will also continue if the element is null. So basically this check is somewhat useless here.

Reply:
0x2000 is 8192 in decimal. Divide this by 32 to get tile coordinates and you get 256.
8192 is indeed the upper limit on sprite coordinates: it means the end of tile 255. (0 = beginning of tile 0, 32 = end of tile 0, beginning of tile 1, etc.)

So, in short: yes, it has.